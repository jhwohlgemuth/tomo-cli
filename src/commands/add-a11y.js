import {
    PackageJsonEditor,
    allDoExist,
    install,
    uninstall
} from '../api';
/**
 * @type {task[]}
 * @see http://pa11y.org/
 */
export const addA11y = [
    {
        text: 'Add accessibility tasks to package.json',
        task: async ({outputDirectory}) => {
            const scripts = {
                'lint:aria': `pa11y ${outputDirectory}/index.html`
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () => allDoExist('package.json')
    },
    {
        text: 'Install pa11y for checking accessibility rules',
        task: ({skipInstall}) => install(['pa11y'], {dev: true, skipInstall}),
        condition: ({skipInstall}) => !skipInstall && allDoExist('package.json')
    }
];
export const removeA11y = [
    {
        text: 'Remove accessibility tasks from package.json',
        task: async () => {
            const scripts = {
                'lint:aria': undefined
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () => allDoExist('package.json')
    },
    {
        text: 'Uninstall pa11y',
        task: () => uninstall(['pa11y']),
        condition: () => allDoExist('package.json') && (new PackageJsonEditor()).hasAll('pa11y')
    }
];
export default addA11y;