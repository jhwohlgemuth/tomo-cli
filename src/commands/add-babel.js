import {
    BabelConfigModuleEditor,
    PackageJsonEditor,
    SnowpackConfigEditor,
    allDoExist,
    allDoNotExist,
    install
} from '../api';

const BABEL_CORE = [
    '@babel/cli',
    '@babel/core',
    '@babel/runtime'
];
const BABEL_PRESETS = [
    '@babel/preset-env',
    'babel-preset-minify'
];
const BABEL_PLUGINS = [
    '@babel/plugin-transform-runtime',
    '@babel/plugin-proposal-class-properties',
    '@babel/plugin-proposal-export-default-from',
    '@babel/plugin-proposal-optional-chaining'
];
const BABEL_REACT_PLUGINS = [
    'react-hot-loader'
];
const BABEL_REACT_PRESETS = [
    '@babel/preset-react'
];
const BABEL_DEPENDENCIES = [
    ...BABEL_CORE,
    ...BABEL_PRESETS,
    ...BABEL_PLUGINS
];
const SNOWPACK_DEPENDENCIES = [
    'snowpack',
    '@snowpack/app-scripts-react',
    '@snowpack/plugin-react-refresh'
];
/**
 * @type {task[]}
 * @see https://babeljs.io/
 */
export const addBabel = [
    {
        text: 'Create Babel config file',
        task: async () => {
            await (new BabelConfigModuleEditor())
                .create()
                .commit();
        },
        condition: () => allDoNotExist('babel.config.js', '.babelrc', '.babelrc.js')
    },
    {
        text: 'Create Snowpack config and Configure Babel for Snowpack',
        task: async ({useSnowpack}) => {
            const snowpackConfig = {
                extends: `'@snowpack/app-scripts-react/babel.config.json'`,
                plugins: `[]`,
                presets: `[]`,
                env: {
                    development: {
                        plugins: `['react-refresh/babel']`
                    }
                }
            };
            await (new SnowpackConfigEditor())
                .create()
                .commit();
            await (new BabelConfigModuleEditor())
                .extend(useSnowpack ? snowpackConfig : {})
                .commit();
        },
        condition: ({useSnowpack}) => useSnowpack && allDoExist('babel.config.js'),
        optional: ({useSnowpack}) => useSnowpack
    },
    {
        text: 'Add React support to Babel configuration file',
        task: async ({useRollup}) => {
            const addQuotes = str => `'${str}'`;
            const maybeRemove = name => (!useRollup || name !== 'react-hot-loader');
            const maybeRename = name => (name === 'react-hot-loader') ? 'react-hot-loader/babel' : name;
            const plugins = [...BABEL_REACT_PLUGINS, ...BABEL_PLUGINS].filter(maybeRemove).map(name => name |> maybeRename |> addQuotes);
            const presets = [...BABEL_PRESETS, ...BABEL_REACT_PRESETS].map(addQuotes);
            await (new BabelConfigModuleEditor())
                .extend({plugins, presets})
                .commit();
        },
        condition: ({useReact, useSnowpack}) => !useSnowpack && useReact && allDoExist('babel.config.js'),
        optional: ({useReact, useSnowpack}) => !useSnowpack && useReact
    },
    {
        text: 'Add Babel build task to package.json',
        task: async ({outputDirectory, sourceDirectory, useSnowpack}) => {
            const scripts = {
                'build:es': `babel ${sourceDirectory} --out-dir ${outputDirectory}`,
                'watch:es': `watch 'npm run build:es' ${sourceDirectory}`
            };
            await (new PackageJsonEditor())
                .extend(useSnowpack ? {scripts: {build: 'snowpack build'}} : {scripts})
                .commit();

        },
        condition: () => allDoExist('package.json')
    },
    {
        text: 'Install Babel core, CLI, presets, and plugins',
        task: ({skipInstall}) => install(BABEL_DEPENDENCIES, {dev: true, skipInstall}),
        condition: ({isNotOffline, skipInstall, useSnowpack}) => !useSnowpack && !skipInstall && isNotOffline && (!(new PackageJsonEditor()).hasAll(...BABEL_DEPENDENCIES) && allDoExist('package.json')), // eslint-disable-line max-len
        optional: ({useSnowpack}) => !useSnowpack
    },
    {
        text: 'Install Babel React presets and plugins',
        task: ({skipInstall}) => install([...BABEL_REACT_PRESETS, ...BABEL_REACT_PLUGINS], {dev: true, skipInstall}),
        condition: ({isNotOffline, skipInstall, useReact, useSnowpack}) => !useSnowpack && !skipInstall && isNotOffline && useReact && allDoExist('package.json'), // eslint-disable-line max-len
        optional: ({useReact, useSnowpack}) => !useSnowpack && useReact
    },
    {
        text: 'Install Snowpack dependencies',
        task: ({skipInstall}) => install(SNOWPACK_DEPENDENCIES, {dev: true, skipInstall}),
        condition: ({isNotOffline, skipInstall, useSnowpack}) => useSnowpack && !skipInstall && isNotOffline && (!(new PackageJsonEditor()).hasAll(...SNOWPACK_DEPENDENCIES) && allDoExist('package.json')), // eslint-disable-line max-len
        optional: ({useSnowpack}) => useSnowpack
    }
];
export default addBabel;