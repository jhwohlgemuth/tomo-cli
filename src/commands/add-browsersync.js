import {
    PackageJsonEditor,
    allDoExist,
    install,
    someDoExistSync,
    uninstall
} from '../api';

const BROWSERSYNC_DEPENDENCIES = [
    'browser-sync',
    'npm-run-all'
];
/**
 * @type {task[]}
 * @see https://www.browsersync.io/docs/command-line
 */
export const addBrowsersync = [
    {
        text: 'Add Browsersync tasks to package.json',
        task: async ({outputDirectory, port}) => {
            const scripts = {
                prestart: 'npm run build:es',
                start: `npm-run-all --parallel watch:es watch:css serve`,
                serve: `browser-sync start --server ${outputDirectory} --files ${outputDirectory} --port ${port}`
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: ({useParcel}) => allDoExist('package.json', 'postcss.config.js') && (someDoExistSync('webpack.config.js', 'rollup.config.js') || useParcel) // eslint-disable-line max-len
    },
    {
        text: 'Install Browsersync dependencies',
        task: ({skipInstall}) => install(BROWSERSYNC_DEPENDENCIES, {dev: true, skipInstall}),
        condition: ({isNotOffline, skipInstall}) => !skipInstall && isNotOffline && allDoExist('package.json')
    }
];
export const removeBrowsersync = [
    {
        text: 'Remove Browsersync tasks from package.json',
        task: async () => {
            const scripts = {
                prestart: undefined,
                start: undefined,
                serve: undefined
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () => allDoExist('package.json')
    },
    {
        text: 'Uninstall Browsersync dependencies',
        task: () => uninstall(BROWSERSYNC_DEPENDENCIES),
        condition: ({skipInstall}) => !skipInstall && allDoExist('package.json') && (new PackageJsonEditor()).hasAll(...BROWSERSYNC_DEPENDENCIES),
        optional: ({skipInstall}) => !skipInstall
    }
];
export default addBrowsersync;