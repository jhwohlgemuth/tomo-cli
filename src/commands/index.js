/* eslint-disable no-magic-numbers */
import {choose, withOptions} from '../utils';
import {
    createPackageJson,
    createSourceDirectory
} from './common';
import addA11y from './add-a11y';
import addBabel from './add-babel';
import {addBrowsersync} from './add-browsersync';
import addEsdoc from './add-esdoc';
import addEslint from './add-eslint';
import addJest from './add-jest';
import addMakefile from './add-makefile';
import addMarionette from './add-marionette';
import {addParcel, removeParcel} from './add-parcel';
import {addPostcss, removePostcss} from './add-postcss';
import {addRollup, removeRollup} from './add-rollup';
import {addWebpack, removeWebpack} from './add-webpack';
import createServer from './create-server';

const createProject = [
    ...createPackageJson,
    ...createSourceDirectory,
    ...addBabel,
    ...addEslint
];
const create = {
    project: createProject,
    app: [
        ...createProject,
        ...addPostcss,
        ...addJest,
        choose({
            default: addMarionette,
            useReact: []// do nothing
        }),
        choose({
            default: addWebpack,
            useRollup: addRollup,
            useParcel: addParcel
        }),
        choose({
            default: addBrowsersync,
            useParcel: []// do nothing
        })
    ],
    server: [
        withOptions({sourceDirectory: '.'}),
        ...createPackageJson,
        ...addEslint,
        ...createServer
    ]
};
const add = {
    a11y: addA11y,
    babel: addBabel,
    browsersync: addBrowsersync,
    esdoc: addEsdoc,
    eslint: [
        ...addBabel,
        ...addEslint
    ],
    jest: [
        ...addBabel,
        ...addJest
    ],
    makefile: addMakefile,
    parcel: [
        ...addBabel,
        ...addParcel
    ],
    postcss: addPostcss,
    rollup: [
        ...addBabel,
        ...addRollup
    ],
    webpack: [
        ...addBabel,
        ...addWebpack
    ]
};
const remove = {
    parcel: removeParcel,
    postcss: removePostcss,
    rollup: removeRollup,
    webpack: removeWebpack
};

export default {
    add,
    remove,
    create,
    new: create// alias for create
};
/**
 * @typedef {Object} task
 * @property {string} text Display text for task
 * @property {function} task Task to execute
 * @property {function} condition Predicate to decide when to execute the task (true) or not (false) - can be async or sync
 * @property {function} [optional] Predicate to decide when to show the task (true) or not (false) - MUST be sync
 */