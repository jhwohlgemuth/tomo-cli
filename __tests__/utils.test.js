import {join} from 'path';
import sinon from 'sinon';
import execa from 'execa';
import {readMakefile as read, useTemporaryDirectory} from './tomo-test';
import commands from '../src/commands';
import {
    EslintConfigModuleEditor,
    MakefileEditor,
    PackageJsonEditor,
    choose,
    createFunctionModuleEditor,
    getElapsedSeconds,
    getElapsedTime,
    getIntendedInput,
    getVersions,
    install,
    uninstall,
    withOptions
} from '../src/api';

jest.mock('execa');

const testDirectory = join(__dirname, 'tomo-fixtures');
/**
 * Verify package.json editor can create and edit a package.json manifest file
 * @test {PackageJsonEditor}
 */
describe('package.json mem-fs editor', () => {
    let pkg;
    beforeEach(() => {
        pkg = new PackageJsonEditor(testDirectory);
    });
    test('create', async () => {
        pkg = new PackageJsonEditor('/some/directory');
        expect(pkg.read()).toEqual('');
        await pkg.create();
        expect(pkg.read()).toMatchSnapshot();
    });
    test('read', () => {
        expect(pkg.read()).toMatchSnapshot();
    });
    test('extend', async () => {
        await pkg.extend({scripts: {foo: 'bar'}});
        expect(pkg.read()).toMatchSnapshot();
    });
    test('copy', async () => {
        const newDirPath = join(testDirectory, 'new');
        await pkg.copy(newDirPath);
        expect(pkg.fs.readJSON(join(newDirPath, 'package.json'))).toMatchSnapshot();
    });
    test('delete', async () => {
        pkg = new PackageJsonEditor('/some/directory');
        expect(pkg.read()).toEqual('');
        await pkg.create();
        expect(pkg.read()).toMatchSnapshot();
        await pkg.delete();
        expect(pkg.read()).toEqual('');
    });
    test('hasSome', () => {
        expect(pkg.hasSome('react')).toBeFalsy();
        expect(pkg.hasSome('webpack', 'execa')).toBeTruthy();
        expect(pkg.hasSome('eslint')).toBeTruthy();
    });
    test('hasAll', () => {
        expect(pkg.hasAll('react')).toBeFalsy();
        expect(pkg.hasAll('some-module', 'execa')).toBeFalsy();
        expect(pkg.hasAll('chalk', 'execa')).toBeTruthy();
        // expect(pkg.hasAll('chalk', 'execa', '@babel/cli')).toBeTruthy();
        expect(pkg.hasAll('eslint')).toBeTruthy();
    });
});
/**
 * Verify ESLint module editor can create and edit an ESLint configuration file
 * @test {EslintConfigModuleEditor}
 */
describe('.eslintrc.js mem-fs editor', () => {
    let cfg;
    let tempDirectory;
    const [setTempDir, cleanupTempDir] = useTemporaryDirectory();
    beforeEach(async () => {
        tempDirectory = await setTempDir();
        process.chdir(tempDirectory);
        cfg = new EslintConfigModuleEditor(testDirectory);
    });
    afterEach(async () => {
        await cleanupTempDir();
    });
    test('create', async () => {
        cfg = new EslintConfigModuleEditor('/some/directory');
        expect(cfg.read()).toEqual('');
        await cfg.create();
        expect(cfg.read()).toMatchSnapshot();
    });
    test('read', () => {
        expect(cfg.read()).toMatchSnapshot();
    });
    test('extend', async () => {
        expect(await cfg.read()).toMatchSnapshot();
        await cfg.extend({key: {foo: `'foo'`}});
        expect(await cfg.read()).toMatchSnapshot();
        await cfg.extend({key: {bar: `'bar'`}});
        expect(await cfg.read()).toMatchSnapshot();
    });
    test('delete', async () => {
        cfg = new EslintConfigModuleEditor('/some/directory');
        expect(cfg.read()).toEqual('');
        await cfg.create();
        expect(cfg.read()).toMatchSnapshot();
        await cfg.delete();
        expect(cfg.read()).toEqual('');
    });
    test('prepend', async () => {
        await cfg.prepend(`const {existsSync} = require('fs-extra');`);
        expect(cfg.read()).toMatchSnapshot();
        await cfg.prepend(`const {join} = require('path');`);
        expect(cfg.read()).toMatchSnapshot();
        await cfg.extend({key: {baz: 'baz'}});
        expect(cfg.read()).toMatchSnapshot();
    });
});
describe('Makefile editor', () => {
    let makefile;
    beforeEach(() => {
        makefile = new MakefileEditor(testDirectory);
    });
    test('create', async () => {
        await makefile.create();
        expect(read(makefile)).toMatchSnapshot();
    });
    test('append', async () => {
        await makefile.append('test line');
        expect(read(makefile)).toMatchSnapshot();
    });
    test('addTask', async () => {
        await makefile
            .addTask('foo', ['echo foo'], {description: 'Foo task'})
            .addTask('bar', ['echo bar'], {description: 'Bar task'})
            .done();
        expect(read(makefile)).toMatchSnapshot();
    });
    test('addComment', async () => {
        await makefile.addComment('Knowledge of the Holy One is understanding');
        expect(read(makefile)).toMatchSnapshot();
    });
    test('importScripts', () => {
        expect(makefile.scripts).toMatchSnapshot();
        makefile.importScripts();
        expect(makefile.scripts).toMatchSnapshot();
    });
    test('appendScripts (no bin variable)', async () => {
        makefile = new MakefileEditor(join(testDirectory, 'other-directory'));
        await makefile
            .importScripts()
            .appendScripts()
            .done();
        expect(read(makefile)).toMatchSnapshot();
    });
    test('appendScripts (with bin variable)', async () => {
        await makefile
            .importScripts()
            .appendScripts()
            .done();
        expect(read(makefile)).toMatchSnapshot();
    });
});
describe('createFunctionModuleEditor', () => {
    const content = {foo: `'bar'`};
    test('can create function module (defaults)', async () => {
        const Module = createFunctionModuleEditor('webpack.config.js', content);
        expect(await (new Module()).create().read()).toMatchSnapshot();
    });
    test('can create function module with passed parameters', async () => {
        const params = ['foo', 'bar'];
        const Module = createFunctionModuleEditor('webpack.config.js', content, {params});
        expect(await (new Module()).create().read()).toMatchSnapshot();
    });
    test('can create function module with passed []', async () => {
        const params = [];
        const Module = createFunctionModuleEditor('webpack.config.js', content, {params});
        expect(await (new Module()).create().read()).toMatchSnapshot();
    });
    test('can create function ES module (defaults)', async () => {
        const esm = true;
        const Module = createFunctionModuleEditor('webpack.config.js', content, {esm});
        expect(await (new Module()).create().read()).toMatchSnapshot();
    });
    test('can create function ES module with passed parameters', async () => {
        const esm = true;
        const params = ['foo', 'bar'];
        const Module = createFunctionModuleEditor('webpack.config.js', content, {esm, params});
        expect(await (new Module()).create().read()).toMatchSnapshot();
    });
    test('can create function ES module with passed []', async () => {
        const esm = true;
        const params = [];
        const Module = createFunctionModuleEditor('webpack.config.js', content, {esm, params});
        expect(await (new Module()).create().read()).toMatchSnapshot();
    });
});
describe('choose via options', () => {
    test('with and without default options', () => {
        const withoutDefault = {
            a: ['a', 'aa', 'aaa'],
            b: ['b', 'bb', 'bbb'],
            c: ['c', 'cc', 'ccc']
        };
        const withDefault = {
            a: ['a', 'aa', 'aaa'],
            b: ['b', 'bb', 'bbb'],
            c: ['c', 'cc', 'ccc'],
            default: [0, 0, 0]
        };
        expect(choose(withoutDefault)({d: true})).toEqual(withoutDefault.a);
        expect(choose(withDefault)({d: true})).toEqual(withDefault.default);
        expect(choose(withDefault)({b: true})).toEqual(withDefault.b);
        expect(choose(withDefault)({b: true, c: true})).toEqual(withDefault.b);
        expect(choose(withDefault)({a: false, c: true})).toEqual(withDefault.c);
    });
});
describe('getElapsedTime', () => {
    const second = 1000;
    const minute = 60 * second; // eslint-disable-line no-magic-numbers
    const clock = sinon.useFakeTimers();
    afterAll(() => {
        clock.restore();
    });
    test('can count seconds', () => {
        const [start] = process.hrtime();
        expect(getElapsedTime(start)).toEqual('00:00:00');
        clock.tick(second);
        expect(getElapsedTime(start)).toEqual('00:00:01');
        clock.tick(second);
        expect(getElapsedTime(start)).toEqual('00:00:02');
    });
    test('can update minutes using modular arithmetic', () => {
        const [start] = process.hrtime();
        expect(getElapsedTime(start)).toEqual('00:00:00');
        clock.tick(59 * second); // eslint-disable-line no-magic-numbers
        expect(getElapsedTime(start)).toEqual('00:00:59');
        clock.tick(2 * second);
        expect(getElapsedTime(start)).toEqual('00:01:01');
    });
    test('can update hours using modular arithmetic', () => {
        const [start] = process.hrtime();
        expect(getElapsedTime(start)).toEqual('00:00:00');
        clock.tick(59 * minute); // eslint-disable-line no-magic-numbers
        expect(getElapsedTime(start)).toEqual('00:59:00');
        clock.tick(2 * minute);
        expect(getElapsedTime(start)).toEqual('01:01:00');
    });
});
describe('getElapsedSeconds', () => {
    test('can convert durations to seconds', () => {
        expect(getElapsedSeconds('00:00:15')).toEqual(15);
        expect(getElapsedSeconds('00:00:59')).toEqual(59);
        expect(getElapsedSeconds('00:01:59')).toEqual(119);
        expect(getElapsedSeconds('00:25:03')).toEqual(1503);
        expect(getElapsedSeconds('10:05:45')).toEqual(36345);
    });
});
describe('getIntendedInput', () => {
    test('return closest matches', () => {
        const {intendedCommand, intendedTerms} = getIntendedInput(commands, 'ad', ['lint']);
        expect(intendedCommand).toEqual('add');
        expect(intendedTerms).toEqual(['eslint']);
    });
});
describe('getVersions', () => {
    test('handle no module name', async () => {
        expect(await getVersions()).toEqual([]);
    });
    test('format response from npm (only allow valid version strings)', async () => {
        const stdout = '1.0.0 ,\n 2.0.0 ,\n not valid ,\n 3.0.0 ';
        execa.mockResolvedValue({stdout});
        expect(await getVersions('some-module-name')).toMatchSnapshot();
    });
});
describe('install', () => {
    const latest = true;
    const skipInstall = true;
    const dev = true;
    test('handle array of string names', async () => {
        expect(await install()).toEqual(['install']);
        expect(await install(['some-module'])).toEqual(['install', 'some-module@latest']);
        expect(await install(['some-module'])).toEqual(['install', 'some-module@latest']);
        expect(await install(['some-module'], {latest: false, skipInstall})).toEqual(['install', 'some-module']);
        expect(await install(['foo', 'bar'], {latest: false, skipInstall})).toEqual(['install', 'foo', 'bar']);
        expect(await install(['some-module'], {dev, skipInstall})).toEqual(['install', 'some-module', '--save-dev']);
        expect(await install(['some-module'], {dev, latest, skipInstall})).toEqual(['install', 'some-module@latest', '--save-dev']);
        expect(await install(['foo', 'bar'], {latest: false, dev, skipInstall})).toEqual(['install', 'foo', 'bar', '--save-dev']);
    });
    test('only allow valid module names', async () => {
        const INVALID_NAME = 'eLaBorAtE-paCkAgE-with-mixed-case';
        expect(await install([INVALID_NAME])).toEqual(['install']);
        expect(await install(['jest', INVALID_NAME])).toEqual(['install', 'jest@latest']);
    });
});
describe('uninstall', () => {
    test('handle array of string names', async () => {
        expect(await uninstall()).toEqual(['uninstall']);
        expect(await uninstall(['some-module'])).toEqual(['uninstall', 'some-module']);
        expect(await uninstall(['some-module', 'foo', 'bar'])).toEqual(['uninstall', 'some-module', 'foo', 'bar']);
    });
    test('only allow valid module names', async () => {
        const INVALID_NAME = 'eLaBorAtE-paCkAgE-with-mixed-case';
        expect(await uninstall([INVALID_NAME])).toEqual(['uninstall']);
        expect(await uninstall(['jest', INVALID_NAME])).toEqual(['uninstall', 'jest']);
    });
});
describe('withOptions', () => {
    test('add custom options', () => {
        const options = {a: 1, b: 2};
        expect(withOptions({a: 0, c: 1})(options)).toMatchSnapshot();
    });
});