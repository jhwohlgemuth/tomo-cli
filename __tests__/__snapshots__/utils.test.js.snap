// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.eslintrc.js mem-fs editor create 1`] = `
"module.exports = {
    env: {
        es6: true,
        jest: true
    },
    extends: [
        'omaha-prime-grade'
    ],
    parser: 'babel-eslint'
};"
`;

exports[`.eslintrc.js mem-fs editor delete 1`] = `
"module.exports = {
    env: {
        es6: true,
        jest: true
    },
    extends: [
        'omaha-prime-grade'
    ],
    parser: 'babel-eslint',
    key: {
        foo: 'foo',
        bar: 'bar'
    }
};"
`;

exports[`.eslintrc.js mem-fs editor extend 1`] = `
"/* eslint-disable no-unused-vars */
const {join} = require('path');
const {existsSync} = require('fs-extra');

module.exports = {
    env: {es6: true, jest: true},
    extends: ['omaha-prime-grade'],
    parser: 'babel-eslint'
};"
`;

exports[`.eslintrc.js mem-fs editor extend 2`] = `
"module.exports = {
    env: {
        es6: true,
        jest: true
    },
    extends: [
        'omaha-prime-grade'
    ],
    parser: 'babel-eslint',
    key: {
        foo: 'foo'
    }
};"
`;

exports[`.eslintrc.js mem-fs editor extend 3`] = `
"module.exports = {
    env: {
        es6: true,
        jest: true
    },
    extends: [
        'omaha-prime-grade'
    ],
    parser: 'babel-eslint',
    key: {
        foo: 'foo',
        bar: 'bar'
    }
};"
`;

exports[`.eslintrc.js mem-fs editor prepend 1`] = `
"const {existsSync} = require('fs-extra');

module.exports = {
    env: {
        es6: true,
        jest: true
    },
    extends: [
        'omaha-prime-grade'
    ],
    parser: 'babel-eslint',
    key: {
        foo: 'foo',
        bar: 'bar'
    }
};"
`;

exports[`.eslintrc.js mem-fs editor prepend 2`] = `
"const {join} = require('path');
const {existsSync} = require('fs-extra');

module.exports = {
    env: {
        es6: true,
        jest: true
    },
    extends: [
        'omaha-prime-grade'
    ],
    parser: 'babel-eslint',
    key: {
        foo: 'foo',
        bar: 'bar'
    }
};"
`;

exports[`.eslintrc.js mem-fs editor prepend 3`] = `
"const {join} = require('path');
const {existsSync} = require('fs-extra');

module.exports = {
    env: {
        es6: true,
        jest: true
    },
    extends: [
        'omaha-prime-grade'
    ],
    parser: 'babel-eslint',
    key: {
        foo: 'foo',
        bar: 'bar',
        baz: baz
    }
};"
`;

exports[`.eslintrc.js mem-fs editor read 1`] = `
"/* eslint-disable no-unused-vars */
const {join} = require('path');
const {existsSync} = require('fs-extra');

module.exports = {
    env: {es6: true, jest: true},
    extends: ['omaha-prime-grade'],
    parser: 'babel-eslint'
};"
`;

exports[`Makefile editor addComment 1`] = `
"# Makefile created with tomo
# Knowledge of the Holy One is understanding"
`;

exports[`Makefile editor addTask 1`] = `
"# Makefile created with tomo
foo: ## Foo task
	echo foo
bar: ## Bar task
	echo bar"
`;

exports[`Makefile editor append 1`] = `
"# Makefile created with tomo
test line"
`;

exports[`Makefile editor appendScripts (no bin variable) 1`] = `
"# Makefile created with tomo


test: ## Task description
	jest --coverage

lint: ## Task description
	eslint . -c ./.eslintrc.js

lint-watch: ## Task description
	echo Starting lint:watch
	watch 'make lint' src

build: ## Task description
	echo Starting build
	webpack
	echo Build complete
"
`;

exports[`Makefile editor appendScripts (with bin variable) 1`] = `
"# Makefile created with tomo
bin := home/user/project/__tests__/tomo-fixtures/node_modules/.bin/


dev: ## Task description
	stmux [ 'make build-watch' : 'make lint-watch' ]

test: ## Task description
	jest --coverage

test-watch: ## Task description
	jest --coverage --watchAll

lint: ## Task description
	$(bin)eslint . -c ./.eslintrc.js

lint-watch: ## Task description
	echo Starting lint:watch
	watch 'make lint' src

build: ## Task description
	echo Starting build
	$(bin)webpack
	echo Build complete

docs: ## Task description
	$(MAKE) lint-docs
	$(MAKE) build-docs
	$(MAKE) open-docs

deploy: ## Task description
	npm-run-all lint build test
"
`;

exports[`Makefile editor create 1`] = `"# Makefile created with tomo"`;

exports[`Makefile editor importScripts 1`] = `Object {}`;

exports[`Makefile editor importScripts 2`] = `
Object {
  "build": "webpack",
  "deploy": "npm-run-all lint build test",
  "dev": "stmux [ \\"npm run build:watch\\" : \\"npm run lint:watch\\" ]",
  "docs": "npm run build:docs",
  "lint": "eslint . -c ./.eslintrc.js",
  "lint:watch": "watch 'npm run lint' src",
  "postbuild": "echo Build complete",
  "postdocs": "npm run open:docs",
  "prebuild": "echo Starting build",
  "predocs": "npm run lint:docs",
  "prelint:watch": "echo Starting lint:watch",
  "test": "jest --coverage",
  "test:watch": "npm test -- --watchAll",
}
`;

exports[`createFunctionModuleEditor can create function ES module (defaults) 1`] = `
"export default () => ({
    foo: 'bar'
});"
`;

exports[`createFunctionModuleEditor can create function ES module with passed [] 1`] = `
"export default () => ({
    foo: 'bar'
});"
`;

exports[`createFunctionModuleEditor can create function ES module with passed parameters 1`] = `
"export default (foo, bar) => ({
    foo: 'bar'
});"
`;

exports[`createFunctionModuleEditor can create function module (defaults) 1`] = `
"module.exports = () => ({
    foo: 'bar'
});"
`;

exports[`createFunctionModuleEditor can create function module with passed [] 1`] = `
"module.exports = () => ({
    foo: 'bar'
});"
`;

exports[`createFunctionModuleEditor can create function module with passed parameters 1`] = `
"module.exports = (foo, bar) => ({
    foo: 'bar'
});"
`;

exports[`getVersions format response from npm (only allow valid version strings) 1`] = `
Array [
  "1.0.0",
  "2.0.0",
  "3.0.0",
]
`;

exports[`package.json mem-fs editor copy 1`] = `
Object {
  "author": "",
  "dependencies": Object {
    "chalk": "2.4.2",
    "delay": "4.2.0",
    "execa": "1.0.0",
  },
  "description": "",
  "devDependencies": Object {
    "@babel/cli": "^7.4.3",
    "@babel/core": "^7.4.3",
    "eslint": "^5.16.0",
    "webpack": "4.30.0",
  },
  "keywords": Array [],
  "license": "ISC",
  "main": "index.js",
  "name": "fixtures",
  "scripts": Object {
    "build": "webpack",
    "deploy": "npm-run-all lint build test",
    "dev": "stmux [ \\"npm run build:watch\\" : \\"npm run lint:watch\\" ]",
    "docs": "npm run build:docs",
    "lint": "eslint . -c ./.eslintrc.js",
    "lint:watch": "watch 'npm run lint' src",
    "postbuild": "echo Build complete",
    "postdocs": "npm run open:docs",
    "prebuild": "echo Starting build",
    "predocs": "npm run lint:docs",
    "prelint:watch": "echo Starting lint:watch",
    "test": "jest --coverage",
    "test:watch": "npm test -- --watchAll",
  },
  "version": "1.0.0",
}
`;

exports[`package.json mem-fs editor create 1`] = `
Object {
  "description": "A super cool app/server/tool/library/widget/thingy",
  "keywords": Array [],
  "license": "MIT",
  "name": "my-project",
  "version": "0.0.0",
}
`;

exports[`package.json mem-fs editor delete 1`] = `
Object {
  "description": "A super cool app/server/tool/library/widget/thingy",
  "keywords": Array [],
  "license": "MIT",
  "name": "my-project",
  "version": "0.0.0",
}
`;

exports[`package.json mem-fs editor extend 1`] = `
Object {
  "author": "",
  "dependencies": Object {
    "chalk": "2.4.2",
    "delay": "4.2.0",
    "execa": "1.0.0",
  },
  "description": "",
  "devDependencies": Object {
    "@babel/cli": "^7.4.3",
    "@babel/core": "^7.4.3",
    "eslint": "^5.16.0",
    "webpack": "4.30.0",
  },
  "keywords": Array [],
  "license": "ISC",
  "main": "index.js",
  "name": "fixtures",
  "scripts": Object {
    "build": "webpack",
    "deploy": "npm-run-all lint build test",
    "dev": "stmux [ \\"npm run build:watch\\" : \\"npm run lint:watch\\" ]",
    "docs": "npm run build:docs",
    "foo": "bar",
    "lint": "eslint . -c ./.eslintrc.js",
    "lint:watch": "watch 'npm run lint' src",
    "postbuild": "echo Build complete",
    "postdocs": "npm run open:docs",
    "prebuild": "echo Starting build",
    "predocs": "npm run lint:docs",
    "prelint:watch": "echo Starting lint:watch",
    "test": "jest --coverage",
    "test:watch": "npm test -- --watchAll",
  },
  "version": "1.0.0",
}
`;

exports[`package.json mem-fs editor read 1`] = `
Object {
  "author": "",
  "dependencies": Object {
    "chalk": "2.4.2",
    "delay": "4.2.0",
    "execa": "1.0.0",
  },
  "description": "",
  "devDependencies": Object {
    "@babel/cli": "^7.4.3",
    "@babel/core": "^7.4.3",
    "eslint": "^5.16.0",
    "webpack": "4.30.0",
  },
  "keywords": Array [],
  "license": "ISC",
  "main": "index.js",
  "name": "fixtures",
  "scripts": Object {
    "build": "webpack",
    "deploy": "npm-run-all lint build test",
    "dev": "stmux [ \\"npm run build:watch\\" : \\"npm run lint:watch\\" ]",
    "docs": "npm run build:docs",
    "lint": "eslint . -c ./.eslintrc.js",
    "lint:watch": "watch 'npm run lint' src",
    "postbuild": "echo Build complete",
    "postdocs": "npm run open:docs",
    "prebuild": "echo Starting build",
    "predocs": "npm run lint:docs",
    "prelint:watch": "echo Starting lint:watch",
    "test": "jest --coverage",
    "test:watch": "npm test -- --watchAll",
  },
  "version": "1.0.0",
}
`;

exports[`withOptions add custom options 1`] = `
Object {
  "a": 0,
  "b": 2,
  "c": 1,
}
`;
